org-code-javabuilder/lib/src/main/java/org/code/javabuilder/PerformanceTracker.java

```diff
@@ -0,0 +1,113 @@
+package org.code.javabuilder;
+
+import static org.code.protocol.LoggerNames.MAIN_LOGGER;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.util.logging.Logger;
+import org.code.protocol.LoggerConstants;
+import org.json.JSONObject;
+
+/**
+ * This is a helper class that allows performance analysis of a variety of key events that occur
+ * while a user's code is compiling and running. This is mostly intended to help analyze performance
+ * metrics on our Lambdas to drive down the execution time for the portions we control, but it can
+ * also be used within our local dev instances of Javabuilder, such as the WebSocketServer.
+ * Throughout the execution of a student's project, events can be logged to the performance tracker.
+ * Then, at the end of the project execution, PerformanceTracker.logPerformance can be invoked in
+ * order to create a single record of the performance of the project execution.
+ */
+public class PerformanceTracker {
+  private static final String FIRST_INSTANCE = "firstInstance";
+  private static final String COLD_BOOT_START = "coldBootStart";
+  private static final String COLD_BOOT_END = "coldBootEnd";
+  private static final String COLD_BOOT_TIME = "coldBootTime";
+  private static final String INSTANCE_START = "instanceStart";
+  private static final String COMPILE_START = "compileStart";;
+  private static final String INITIALIZATION_TIME = "initializationTime";
+  private static final String COMPILE_END = "compileEnd";
+  private static final String USER_CODE_START = "userCodeStart";
+  private static final String TRANSITION_TIME = "transitionTime";
+  private static final String USER_CODE_END = "userCodeEnd";
+  private static final String INSTANCE_END = "instanceEnd";
+  private static final String CLEANUP_TIME = "cleanupTime";
+  private static final String TYPE = "performanceReport";
+  private final JSONObject logs;
+  private static PerformanceTracker instance;
+
+  private PerformanceTracker() {
+    logs = new JSONObject();
+  }
+
+  public static PerformanceTracker getInstance() {
+    if (PerformanceTracker.instance == null || PerformanceTracker.instance.logs == null) {
```

Comment:
Good point. I don't think it would be null. I included this here after accidentally setting logs to null in a different location (I must not have had it marked final at that point yet). Are you thinking that bit should be removed?

Reply:
Yeah, I think it should be safe to remove since in theory there shouldn't be a situation in which `logs` get set to null, unless we want to guard against future changes?